src/dbnode/persist/fs/wide/entry_checksum_mismatch_checker_test.go

```diff
@@ -0,0 +1,376 @@
+		Checksums: []int64{1, 2},
+		EndMarker: []byte("foo2"),
+	})
+
+	chk := NewEntryChecksumMismatchChecker(reader, buildOpts(t))
+	_, err := chk.ComputeMismatchForEntry(idxEntry("bar2", "bar"))
+	require.Error(t, err)
+}
+
+func assertNoMismatch(
+	t *testing.T,
+	chk EntryChecksumMismatchChecker,
+	entry schema.IndexEntry,
+) {
+	mismatch, err := chk.ComputeMismatchForEntry(entry)
+	require.NoError(t, err)
+	assert.Equal(t, 0, len(mismatch))
+}
+
+func TestComputeMismatchWithDelayedReader(t *testing.T) {
+	ch := make(chan ident.IndexChecksumBlock)
+	reader := NewIndexChecksumBlockReader(ch)
+	chk := NewEntryChecksumMismatchChecker(reader, buildOpts(t))
+
+	go func() {
+		time.Sleep(time.Millisecond * 100)
+		ch <- ident.IndexChecksumBlock{
+			Checksums: []int64{1},
+			EndMarker: []byte("foo1"),
+		}
+		time.Sleep(time.Millisecond * 200)
+		ch <- ident.IndexChecksumBlock{
+			Checksums: []int64{10},
+			EndMarker: []byte("qux10"),
+		}
+		close(ch)
+	}()
+
+	assertNoMismatch(t, chk, idxEntry("foo1", "bar"))
+	assertNoMismatch(t, chk, idxEntry("qux10", "baz"))
+	assert.Equal(t, 0, len(chk.Drain()))
+}
+
+func TestComputeMismatchNoMismatch(t *testing.T) {
+	reader := buildTestReader(ident.IndexChecksumBlock{
+		Checksums: []int64{1, 2, 3},
+		EndMarker: []byte("foo3"),
+	}, ident.IndexChecksumBlock{
+		Checksums: []int64{100, 5},
+		EndMarker: []byte("zoo5"),
+	})
+
+	chk := NewEntryChecksumMismatchChecker(reader, buildOpts(t))
+	assertNoMismatch(t, chk, idxEntry("abc1", "aaa"))
+	assertNoMismatch(t, chk, idxEntry("def2", "bbb"))
+	assertNoMismatch(t, chk, idxEntry("foo3", "ccc"))
+	assertNoMismatch(t, chk, idxEntry("qux100", "ddd"))
+	assertNoMismatch(t, chk, idxEntry("zoo5", "eee"))
+	assert.Equal(t, 0, len(chk.Drain()))
+}
+
+func TestComputeMismatchMismatchesPrimary(t *testing.T) {
+	reader := buildTestReader(ident.IndexChecksumBlock{
+		Checksums: []int64{1, 2, 3},
+		EndMarker: []byte("foo3"),
+	}, ident.IndexChecksumBlock{
+		Checksums: []int64{4, 5},
+		EndMarker: []byte("moo5"),
+	}, ident.IndexChecksumBlock{
+		Checksums: []int64{6, 7, 8},
+		EndMarker: []byte("qux8"),
+	}, ident.IndexChecksumBlock{
+		Checksums: []int64{9, 10},
+		EndMarker: []byte("zzz9"),
+	})
+
+	chk := NewEntryChecksumMismatchChecker(reader, buildOpts(t))
+
+	expected := []ReadMismatch{
+		testIdxMismatch(1),
+		testIdxMismatch(2),
+	}
+
+	mismatches, err := chk.ComputeMismatchForEntry(idxEntry("foo3", "ccc"))
+	require.NoError(t, err)
+	testMismatches(t, expected, mismatches)
+
+	expected = []ReadMismatch{
+		testIdxMismatch(4),
+		testIdxMismatch(5),
+		testIdxMismatch(6),
+	}
+
+	mismatches, err = chk.ComputeMismatchForEntry(idxEntry("qux7", "ddd"))
+	require.NoError(t, err)
+	testMismatches(t, expected, mismatches)
+
+	expected = []ReadMismatch{
+		testIdxMismatch(8),
+		testIdxMismatch(9),
+		testIdxMismatch(10),
+	}
+	testMismatches(t, expected, chk.Drain())
+}
+
+func TestComputeMismatchMismatchesSecondary(t *testing.T) {
```

Comment:
I am obviously missing something, but how do we indicate primary vs secondary mismatches? Maybe something in the naming could explicitly indicate that?

Reply:
Yeah, it's more of index checksum mismatch vs entry mismatch semantics now; will update